4.2.0 won't compile on AIX 7.1

zeromq 4.1.6 compiles on my AIX box and passes all tests. However, attempting to build zeromq 4.2.0 I find compile problems. Details follow:

$ xlC -qversion
IBM XL C/C++ for AIX, V12.1 (5765-J02, 5725-C72)
Version: 12.01.0000.0000
$ oslevel
7.1.0.0

$ configure CC=xlc CXX=xlC
$ make
Making all in doc
make[1]: Entering directory `/home/meb/zeromq-4.2.0/doc'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/home/meb/zeromq-4.2.0/doc'
make[1]: Entering directory `/home/meb/zeromq-4.2.0'
  CXX      src/src_libzmq_la-proxy.lo
"src/proxy.cpp", line 157.34: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
"src/proxy.cpp", line 188.22: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
"src/proxy.cpp", line 189.51: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
"src/proxy.cpp", line 197.22: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
"src/proxy.cpp", line 198.25: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
make[1]: *** [src/src_libzmq_la-proxy.lo] Error 1
make[1]: Leaving directory `/home/meb/zeromq-4.2.0'
make: *** [all-recursive] Error 1

I get as far as

make[1]: Entering directory `/home/meb/libzmq-aix_poll'
  CXX    src/src_libzmq_la-proxy.lo
"src/proxy.cpp", line 159.34: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
"src/proxy.cpp", line 190.22: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
"src/proxy.cpp", line 191.51: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
"src/proxy.cpp", line 199.22: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
"src/proxy.cpp", line 200.25: 1540-0217 (S) "rtnevents" is not a member of "struct zmq_pollitem_t".
make[1]: *** [src/src_libzmq_la-proxy.lo] Error 1
make[1]: Leaving directory `/home/meb/libzmq-aix_poll'
make: *** [all-recursive] Error 1

So this is the same problem (but different line numbers).

However, interestingly, if I export OBJECT_MODE=64 (to build 64 bit code) then the code compiles and the library is built, but ALL of the tests fail with the same error message, reproduced here for one test:

Bad file number (src/signaler.cpp:162)
/bin/sh: 1312030 IOT/Abort trap(coredump)
FAIL: tests/test_system

I think it might be useful to show the relevant contents of the AIX <sys/poll.h> header file? Here is the portion that might be a source of headaches:

…
#ifdef  __64BIT__
/* pollfd.fd is always a 32 bit entity (XOPEN) */
struct  pollfd
{
        int     fd;                     /* file descriptor or file ptr  */
        short   events;                 /* requested events             */
        short   revents;                /* returned events              */

};
#else /* __64BIT__ */
struct  pollfd
{
        long    fd;                     /* file descriptor or file ptr  */
        ushort  reqevents;              /* requested events             */
        ushort  rtnevents;              /* returned events              */
};
#ifndef __HIDE_SVR4_POLLFD_NAMES
#define events  reqevents               /* SVR3,4 pollfd member name    */
#define revents rtnevents               /* SVR3,4 pollfd member name    */
#endif /* __HIDE_SVR4_POLLFD_NAMES */
#endif /* __64BIT__ */
…



From: Luca Boccassi [mailto:notifications@github.com]
Sent: Thursday, December 01, 2016 4:41 PM
To: zeromq/libzmq
Cc: Mike Beddo; Author
Subject: Re: [zeromq/libzmq] 4.2.0 won't compile on AIX 7.1 (#2234)


Since I don't have access to an AIX system, would you be able to test this fix?

https://github.com/bluca/libzmq/tree/aix_poll

The include order was changed at some point, I suspect that's the issue.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/zeromq/libzmq/issues/2234#issuecomment-264330191>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AUEOrNsy7OQfcgCC6wVNjDlmGG9qr42pks5rD1sPgaJpZM4LBvXc>.

Using the source code I’ve been working with you on, I run the test and this is what I got.

$ LD_LIBRARY_PATH=src/.libs tests/.libs/test_pair_tcp
Bad file number (src/signaler.cpp:162)
IOT/Abort trap(coredump)
$

I next tried the aix_poll branch you pointed me to. At first, configure complained about missing libsodium, so I reran with

$ configure CC=xlc CXX=xlC –without-libsodium

The library made with a single warning:

  CXX    src/src_libzmq_la-socket_base.lo
"src/socket_base.cpp", line 162.26: 1540-2411 (W) A dynamic cast is present, but the correct RTTI option is not specified.

I tried a “make check” but got this:

  CXX    tests/test_stream_exceeds_buffer.o
"tests/test_stream_exceeds_buffer.cpp", line 16.30: 1540-0274 (S) The name lookup for "AF_INET" did not find a declaration.
"tests/test_stream_exceeds_buffer.cpp", line 16.39: 1540-0274 (S) The name lookup for "SOCK_STREAM" did not find a declaration.
"tests/test_stream_exceeds_buffer.cpp", line 19.38: 1540-0274 (S) The name lookup for "SOL_SOCKET" did not find a declaration.
"tests/test_stream_exceeds_buffer.cpp", line 19.50: 1540-0274 (S) The name lookup for "SO_REUSEADDR" did not find a declaration.
"tests/test_stream_exceeds_buffer.cpp", line 28.10: 1540-0274 (S) The name lookup for "bind" did not find a declaration.
"tests/test_stream_exceeds_buffer.cpp", line 30.10: 1540-0274 (S) The name lookup for "listen" did not find a declaration.
"tests/test_stream_exceeds_buffer.cpp", line 40.23: 1540-0274 (S) The name lookup for "accept" did not find a declaration.
"tests/test_stream_exceeds_buffer.cpp", line 46.10: 1540-0274 (S) The name lookup for "send" did not find a declaration.
make[2]: *** [tests/test_stream_exceeds_buffer.o] Error 1
make[2]: Leaving directory `/home/meb/libzmq'
make[1]: *** [check-am] Error 2
make[1]: Leaving directory `/home/meb/libzmq'
make: *** [check-recursive] Error 1

But since the “make check” built test_pair_tcp I decided to try that:

$ LD_LIBRARY_PATH=src/.libs tests/.libs/test_pair_tcp
$

So it seemed to work (?).

From: Luca Boccassi [mailto:notifications@github.com]
Sent: Monday, December 05, 2016 4:04 PM
To: zeromq/libzmq
Cc: Mike Beddo; Author
Subject: Re: [zeromq/libzmq] 4.2.0 won't compile on AIX 7.1 (#2234)


Could you please run one of the tests manually to see what's the precise error?

LD_LIBRARY_PATH=src/.libs tests/.libs/test_pair_tcp

(This works on Linux, I hope it works on AIX too!)

Also, could you please do one more try with this branch and see if it still compiles? If it does I'll open a PR so that we can make progress. Once it at least builds then we can continue with the test case failures.

https://github.com/bluca/libzmq/tree/aix_poll

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/zeromq/libzmq/issues/2234#issuecomment-265007003>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AUEOrG7lTecFTr5rHOfxzv3BwxiP7Scjks5rFJhMgaJpZM4LBvXc>.

--- a/include/zmq.h
+++ b/include/zmq.h
@@ -112,6 +112,13 @@ extern "C" {
 #   include <stdint.h>
 #endif
 
+//  32-bit AIX's pollfd struct members are called reqevents and rtnevents so it
+//  defines compatibility macros for them. Need to include that header first to
+//  stop build failures since zmq_pollset_t defines them as events and revents.
+#ifdef ZMQ_HAVE_AIX
+    #include <poll.h>
+#endif
+
 
 /******************************************************************************/
 /*  0MQ errors.                                                               */

--- a/src/proxy.cpp
+++ b/src/proxy.cpp
@@ -27,17 +27,23 @@
     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 
+//  On AIX platform, poll.h has to be included first to get consistent
+//  definition of pollfd structure (AIX uses 'reqevents' and 'retnevents'
+//  instead of 'events' and 'revents' and defines macros to map from POSIX-y
+//  names to AIX-specific names).
+//  zmq.h must be included *after* poll.h for AIX to build properly.
+//  precompiled.hpp includes include/zmq.h
+#if defined ZMQ_POLL_BASED_ON_POLL && defined ZMQ_HAVE_AIX
+#include <poll.h>
+#endif
+
 #include "precompiled.hpp"
 #include <stddef.h>
 #include "poller.hpp"
 #include "proxy.hpp"
 #include "likely.hpp"
 
-//  On AIX platform, poll.h has to be included first to get consistent
-//  definition of pollfd structure (AIX uses 'reqevents' and 'retnevents'
-//  instead of 'events' and 'revents' and defines macros to map from POSIX-y
-//  names to AIX-specific names).
-#if defined ZMQ_POLL_BASED_ON_POLL && !defined ZMQ_HAVE_WINDOWS
+#if defined ZMQ_POLL_BASED_ON_POLL && !defined ZMQ_HAVE_WINDOWS && !defined ZMQ_HAVE_AIX
 #include <poll.h>
 #endif
 
@@ -45,10 +51,6 @@
 // dependency chain
 #include "socket_base.hpp"
 #include "err.hpp"
-
-// TODO: determine if this is an issue, since zmq.h is being loaded from pch.
-// zmq.h must be included *after* poll.h for AIX to build properly
-//#include "../include/zmq.h"
 
 int capture(
         class zmq::socket_base_t *capture_,

--- a/src/signaler.cpp
+++ b/src/signaler.cpp
@@ -27,15 +27,21 @@
     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#include "precompiled.hpp"
-#include "poller.hpp"
-
 //  On AIX, poll.h has to be included before zmq.h to get consistent
 //  definition of pollfd structure (AIX uses 'reqevents' and 'retnevents'
 //  instead of 'events' and 'revents' and defines macros to map from POSIX-y
 //  names to AIX-specific names).
+//  zmq.h must be included *after* poll.h for AIX to build properly.
+//  precompiled.hpp includes include/zmq.h
+#if defined ZMQ_POLL_BASED_ON_POLL && defined ZMQ_HAVE_AIX
+#include <poll.h>
+#endif
+
+#include "precompiled.hpp"
+#include "poller.hpp"
+
 #if defined ZMQ_POLL_BASED_ON_POLL
-#if !defined ZMQ_HAVE_WINDOWS
+#if !defined ZMQ_HAVE_WINDOWS && !defined ZMQ_HAVE_AIX
 #include <poll.h>
 #endif
 #elif defined ZMQ_POLL_BASED_ON_SELECT

